### DISCUSSION TITLE
Feature request: Please, add face restoration for Desktop version of Upscayl.

### DISCUSSION THREAD

**Discussion by JohnClaw** [upscayl/upscayl#475] [Ideas]
...

**Comment by NayamAmarshe** [comment]
Not possible for now, sorry. We can't integrate something like that easily in a desktop app for all platforms, it's a HUGE task for a team of 2 developers.

**Reply by etlweather** [reply]
Ah, two devs - I might start helping then!

**Reply by NayamAmarshe** [reply]
It'd still be extremely difficult to get it working. Dealing with python scripts, installing modules correctly, setting up the dev environment on users' machines is a lot of work and can be buggy or simply might not work for most.

**Comment by JohnClaw** [comment]
> It'd still be extremely difficult to get it working. Dealing with python scripts, installing modules correctly, setting up the dev environment on users' machines is a lot of work and can be buggy or simply might not work for most.

And what about c++ versions of face restorators? Do they exist?

**Reply by JohnClaw** [reply]
I found at least one c++ restorator: https://github.com/dianhsu/swin-transformer-cpp

**Comment by JohnClaw** [comment]
Also there are ONNX codeformer and GFPGAN models: https://github.com/harisreedhar/Face-Upscalers-ONNX/releases/tag/Models
CPP resources to implement ONNX runtime:
Microsoft C++ ONNX Runtime for CPU: https://www.nuget.org/packages/Microsoft.ML.OnnxRuntime
https://github.com/hpc203/CodeFormer-onnxrun-cpp-py
https://github.com/xmba15/onnx_runtime_cpp

**Reply by aaronliu0130** [reply]
Wow! We have a way to convert ONNX to our format. I’ll check if these are viable.

**Reply by NayamAmarshe** [reply]
ONNX does not really mean anything because it's just a model format. The code has to be ESRGAN compatible to really work with the binary, then it doesn't really matter if it's PyTorch or ONNX or NCNN because it's guaranteed to run.

**Reply by aaronliu0130** [reply]
Some downloads are prefixed with GFPGAN, which I'm pretty sure is ESRGAN compatible.

**Reply by aaronliu0130** [reply]
After running onnxsim on GFPGAN-1.4, I got this mysterious error:
![image](https://github.com/upscayl/upscayl/assets/11874211/ec58fea1-0ebc-40b8-8ea8-cf11fe8cbebf)
Guess that's all, folks

**Reply by JohnClaw** [reply]
Well, i found another GFPGAN onnx model, try it, please:

https://huggingface.co/facefusion/models/resolve/main/GFPGANv1.4.onnx

**Reply by JohnClaw** [reply]
Maybe this tool will help to convert ONNX to NCNN:

https://github.com/Tencent/ncnn/blob/master/tools/onnx/onnx2ncnn.cpp

**Reply by JohnClaw** [reply]
This site also can convert ONNX to NCNN:

https://convertmodel.com/

**Reply by aaronliu0130** [reply]
> [huggingface.co/facefusion/models/resolve/main/GFPGANv1.4.onnx](https://huggingface.co/facefusion/models/resolve/main/GFPGANv1.4.onnx)

I'm pretty sure that's the exact same model I tried to convert

> https://convertmodel.com/

![image](https://github.com/upscayl/upscayl/assets/11874211/bac6e761-fdd2-454f-b4be-5e25da5952cc)

> https://github.com/Tencent/ncnn/blob/master/tools/onnx/onnx2ncnn.cpp

![image](https://github.com/upscayl/upscayl/assets/11874211/ea9f81ac-348c-459d-ba1f-89dce1e3c2d3)

**Reply by JohnClaw** [reply]
I'm just attempting to find GFPGAN.cpp.exe or Codeformer.cpp.exe or any other CLI.exe that can do face restoration. Planning to use it with images generated in Stable-diffusion.cpp.exe (https://github.com/leejet/stable-diffusion.cpp) and zoomed in Upscayl. Unfortunately, my searchings for such app are unsuccessful.

**Reply by NayamAmarshe** [reply]
> I'm just attempting to find GFPGAN.cpp.exe or Codeformer.cpp.exe or any other CLI.exe that can do face restoration. Planning to use it with images generated in Stable-diffusion.cpp.exe (https://github.com/leejet/stable-diffusion.cpp) and zoomed in Upscayl. Unfortunately, my searchings for such app are unsuccessful.

Yep, it's not an easy task to create such things. Even the ncnn vulkan project was created by a senior AI researcher.

**Comment by JohnClaw** [comment]
> I'm pretty sure that's the exact same model I tried to convert

Ok. I don't lose hope and found some more variants that may solve the problem:

https://github.com/magicse/GFPGANv1.3-to-ncnn
https://github.com/FeiGeChuanShu/GFPGAN-ncnn
https://github.com/onuralpszr/GFPGAN-ncnn-vulkan
https://github.com/Qengineering/GFPGAN-ncnn-Raspberry-Pi-4

**Comment by RafaelLinux** [comment]
I came here precisely because I'm an Upscayl user, but there are cases where, in the default models, faces end up being "restored" into grotesque faces. I thought adding the [GFPGAN](https://github.com/TencentARC/GFPGAN) model would be easy, but the first thing I found was this discussion thread, abandoned since 2023... Has anyone found a way to use GFPGAN from within Upscayl?
